File: Code/Chapter 11/C#/AccChanges/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Data.SqlClient;

namespace AccChanges
{
   public partial class Form1 : Form
   {
      private static string connectionString = "Server=(local);Database=Test;Integrated Security=SSPI";
      private DataSet ds = null;

      public Form1()
      {
         InitializeComponent();
      }

      private void button1_Click(object sender, EventArgs e)
      {
         using (SqlConnection testConnection = new SqlConnection(connectionString))
         {
            SqlDataAdapter sqlDA = new SqlDataAdapter("Select * from Customers",testConnection);
            SqlCommand myUpdateCommand = new SqlCommand("Update Customers Set FirstName = @FirstName where CustomerID = @CustomerID");

            SqlParameter param1 = new SqlParameter("@FirstName", SqlDbType.VarChar, 50);
            param1.SourceColumn = "FirstName";

            SqlParameter param2 = new SqlParameter("@CustomerID", SqlDbType.VarChar, 50);
            param2.SourceColumn = "CustomerID";

            myUpdateCommand.Parameters.AddRange(new SqlParameter[] {param1,param2}) ;
            myUpdateCommand.Connection = testConnection;

            sqlDA.UpdateCommand = myUpdateCommand;

            testConnection.Open();
            SqlTransaction myTransaction = testConnection.BeginTransaction();
            sqlDA.UpdateCommand.Transaction = myTransaction;
            sqlDA.AcceptChangesDuringUpdate = false;

            try
            {
               sqlDA.Update(ds);
               myTransaction.Commit();
               ds.AcceptChanges();
            }
            catch (Exception ex)
            {
               MessageBox.Show(ex.ToString());
            }
         }
      }

      private void Form1_Load(object sender, EventArgs e)
      {
         using (SqlConnection testConnection = new SqlConnection(connectionString))
         {
            SqlDataAdapter sqlDA = new SqlDataAdapter("Select * from Customers", testConnection);
            ds = new DataSet();
            sqlDA.Fill(ds);
            dataGridView1.DataSource = ds.Tables[0];
         }
      }
   }
}

File: Code/Chapter 11/VB.NET/AccChanges/Form1.vb
Imports System.Data
Imports System.Data.SqlClient

Public Class Form1

   Private Shared connectionString As String = "Server=(local);Database=Test;Integrated Security=SSPI"
   Private ds As DataSet = Nothing

   Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
      Using testConnection As SqlConnection = New SqlConnection(connectionString)
         Dim sqlDA As SqlDataAdapter = New SqlDataAdapter("Select * from Customers", testConnection)
         ds = New DataSet()
         sqlDA.Fill(ds)
         dataGridView1.DataSource = ds.Tables(0)
      End Using
   End Sub

   Private Sub button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles button1.Click
      Using testConnection As SqlConnection = New SqlConnection(connectionString)
         Dim sqlDA As SqlDataAdapter = New SqlDataAdapter("Select * from Customers", testConnection)
         Dim myUpdateCommand As SqlCommand = New SqlCommand("Update Customers Set FirstName = @FirstName where CustomerID = @CustomerID")

         Dim param1 As SqlParameter = New SqlParameter("@FirstName", SqlDbType.VarChar, 50)
         param1.SourceColumn = "FirstName"

         Dim param2 As SqlParameter = New SqlParameter("@CustomerID", SqlDbType.VarChar, 50)
         param2.SourceColumn = "CustomerID"

         myUpdateCommand.Parameters.AddRange(New SqlParameter() {param1, param2})
         myUpdateCommand.Connection = testConnection

         sqlDA.UpdateCommand = myUpdateCommand

         testConnection.Open()
         Dim myTransaction As SqlTransaction = testConnection.BeginTransaction()
         sqlDA.UpdateCommand.Transaction = myTransaction
         sqlDA.AcceptChangesDuringUpdate = False

         Try
            sqlDA.Update(ds)
            myTransaction.Commit()
            ds.AcceptChanges()
         Catch ex As Exception
            MessageBox.Show(ex.ToString())
         End Try
      End Using
   End Sub
End Class
